python/lsst/ts/criopy/gui/sal/replay_widget.py

```diff
@@ -0,0 +1,311 @@
+        start = self.start.dateTime()
+        if start <= next_time <= start.addMSecs(int(self.duration.value() * 1000)):
+            self.current.setDateTime(next_time)
+
+    @Slot()
+    def pause(self, checked: bool) -> None:
+        if self.play_timer is not None:
+            self.killTimer(self.play_timer)
+            self.play_timer = None
+
+    @Slot()
+    def step_forward(self, checked: bool) -> None:
+        next_time = self.current.dateTime().addMSecs(self.step_size_box.value())
+        start = self.start.dateTime()
+        if start <= next_time <= start.addMSecs(int(self.duration.value() * 1000)):
+            self.current.setDateTime(next_time)
+
+    @Slot()
+    def play(self, checked: bool) -> None:
+        self.__change_timer(10)
+
+    @Slot()
+    def forward(self, checked: bool) -> None:
+        self.__change_timer(500)
+
+    @Slot()
+    def retrieve_data(self) -> None:
+        self.slider.setMinimum(0)
+        self.slider.setMaximum(self.duration.value() * 1000)
+        self.slider.setValue(0)
+        self.slider_moved(0)
+        self.play_time = self.start.dateTime()
+
+    @Slot()
+    def slider_moved(self, value: int) -> None:
+        new = self.start.dateTime().addMSecs(value)
+        if new != self.current.dateTime():
+            self.current.setDateTime(new)
+
+    @Slot()
+    def download_started(self) -> None:
+        self.setEnabled(False)
+
+    @Slot()
+    def download_finished(self) -> None:
+        self.setEnabled(True)
+
+    @asyncSlot()
+    async def replay(self, date_time: QDateTime) -> None:
+        if self.player is None:
+            self.player = Player(self.sal, self.efd)
+            self.player.downloadStarted.connect(self.download_started)
+            self.player.downloadFinished.connect(self.download_finished)
+
+            self.sal.freeze(self.player.cache)
+
+        timepoint = (
+            date_time.toMSecsSinceEpoch() - self.start.dateTime().toMSecsSinceEpoch()
+        )
+        if timepoint < 0:
+            date_time = self.start.dateTime()
+            timepoint = 0
+        elif timepoint > self.slider.maximum():
+            date_time = self.start.dateTime.addMSecs(int(self.duration.value() * 1000))
+            timepoint = self.slider.maximum()
+
+        self.slider.setValue(timepoint)
+        start = Time(date_time.toMSecsSinceEpoch() / 1000.0, format="unix")
+        await self.player.replay(start, TimeDelta(self.duration.value(), format="sec"))
+
+
+class ReplayWidget(QWidget):
+    """Widget containing control for data replay and a GUI log widget.
+
+    Parameters
+    ----------
+    sal : MetaSAL
+        SAL CSC connection.
+    """
+
+    def __init__(self, sal: MetaSAL):
+        super().__init__()
+        self.setWindowTitle("Replay")
+        self.app_name = "LSST.TS.REPLAY"
+
+        layout = QVBoxLayout()
+        self.setLayout(layout)
+
+        self.replay_control = ReplayControlWidget(sal, "usdf_efd")
```

Comment:
Since this would be performed on the summit, why not use "summit_efd"?

Reply:
OSW-927, added TODO entry.